src/gui/window.rs

```diff
@@ -47,82 +52,250 @@ impl Fps {
-    midi_file: MIDIFileUnion,
+    midi_file: Option<MIDIFileUnion>,
     fps: Fps,
 }
 
 impl GuiWasabiWindow {
     pub fn new(renderer: &mut GuiRenderer) -> GuiWasabiWindow {
-        let mut midi_file = MIDIFileUnion::InRam(InRamMIDIFile::load_from_file(
-            "D:/Midis/The Quarantine Project.mid",
-            SimpleTemporaryPlayer::new(),
-        ));
-
-        midi_file.timer_mut().play();
-
         GuiWasabiWindow {
             render_scene: GuiRenderScene::new(renderer),
             keyboard_layout: keyboard_layout::KeyboardLayout::new(&Default::default()),
             keyboard: GuiKeyboard::new(),
-            midi_file,
+            midi_file: None,
             fps: Fps::new(),
         }
     }
 
     /// Defines the layout of our UI
-    pub fn layout(&mut self, state: &mut GuiState) {
+    pub fn layout(
+        &mut self,
+        state: &mut GuiState,
+        perm_settings: &mut WasabiPermanentSettings,
+        temp_settings: &mut WasabiTemporarySettings,
+    ) {
         let ctx = state.gui.context();
-
+        let window_size = vec![ctx.available_rect().width(), ctx.available_rect().height()];
         self.fps.update();
-
         ctx.set_visuals(Visuals::dark());
+        let note_speed = perm_settings.note_speed;
 
-        // Render the top panel
-        egui::TopBottomPanel::top("Top panel")
-            .height_range(100.0..=100.0)
-            .show(&ctx, |ui| {
-                if let Some(length) = self.midi_file.midi_length() {
-                    let time = self.midi_file.timer().get_time().as_secs_f64();
-                    let progress = (time / length) as f32;
-                    ui.add(egui::ProgressBar::new(progress));
-                }
+        // Render the settings window if the value from
+        // the temporary settings allows it
+        if temp_settings.settings_visible {
+            egui::Window::new("Settings")
+                .resizable(true)
+                .collapsible(true)
+                .title_bar(true)
+                .scroll2([false, true])
+                .enabled(true)
+                .open(&mut temp_settings.settings_visible)
+                .show(&ctx, |ui| {
+                    egui::Grid::new("settings_grid")
+                        .num_columns(2)
+                        .spacing([40.0, 4.0])
+                        .striped(true)
+                        .show(ui, |ui| {
+                            ui.label("SFZ Path: ");
+                            ui.horizontal(|ui| {
+                                ui.add(egui::TextEdit::singleline(&mut perm_settings.sfz_path));
+                                if ui.button("Browse...").clicked() {
+                                    let midi_path: String = FileDialog::new()
+                                        .add_filter("sfz", &["sfz"])
+                                        .set_directory("/")
+                                        .pick_file()
+                                        .unwrap()
```

Comment:
I was testing stuff and wasabi completely crashed when I closed the file picker without picking a file. I noticed it's caused by this unwrap. tl;dr unwraps are bad, both of them.

Reply:
Oh right I added a check when selecting the MIDI but forgot to copy it to the SFZ selection.